node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -541,13 +465,15 @@ async fn declare<Context>(ctx: &mut Context, state: &mut State, peer: PeerId) {
 	}
 }
 
-/// Issue a connection request to a set of validators and
-/// revoke the previous connection request.
+/// Updates a set of connected validators based on their advertisement-bits
+/// in a validators buffer.
 #[overseer::contextbounds(CollatorProtocol, prefix = self::overseer)]
 async fn connect_to_validators<Context>(
 	ctx: &mut Context,
-	validator_ids: Vec<AuthorityDiscoveryId>,
+	validator_groups_buf: &ValidatorGroupsBuffer,
 ) {
+	let validator_ids = validator_groups_buf.validators_to_connect();
```

Comment:
`connect_to_validators` is actually only ever called on `distribute_collation`. This means while we perfectly keep track to whom we should stay connected, we don't make any use of it (as we will only update on the next collation we receive, which is way too late).

Reply:
There is actually already a separate issue about this: https://github.com/paritytech/polkadot/issues/5062 ... but it kind of has to be fixed with this PR, otherwise it won't improve anything.

Reply:
See https://github.com/paritytech/polkadot/pull/6022#issuecomment-1250214286

What this PR actually fixes is keeping connections to multiple groups alive on rotation boundaries, not clearing it up with previous or next group like it was done before.

We could ensure that we call `connect_to_validators` every (e.g.) 15 seconds or something, but I don't see much point in doing so.

Reply:
By the way, the [test](https://github.com/paritytech/polkadot/blob/eab4fefd183125ca82e475635d721933636b9845/node/network/collator-protocol/src/collator_side/tests.rs#L994-L1093) very well expresses which behavior this PR is supposed to fix.

Reply:
Fair, but still - without that fix we actually don't ever disconnect at all (except on the next collation). We could have achieved that with less code. :grimacing: Anyhow, you are right - we can leave that for another PR.

Reply:
e7fac145cc13e29e8b6e6a835b581874ae62d5b6